File: src/ObjectManager/Object.Tes/FilePacks/Records/040-SBSP.Subspace.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class SBSPRecord : Record
    {
        public struct DNAMField
        {
            public float X; // X dimension
            public float Y; // Y dimension
            public float Z; // Z dimension

            public DNAMField(UnityBinaryReader r, int dataSize)
            {
                X = r.ReadLESingle();
                Y = r.ReadLESingle();
                Z = r.ReadLESingle();
            }
        }

        public override string ToString() => $"SBSP: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public DNAMField DNAM;

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "DNAM": DNAM = new DNAMField(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/040-SBSP.Subspace.swift
//
//  SBSPRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class SBSPRecord: Record {
    public struct DNAMField {
        public let x: Float // X dimension
        public let y: Float // Y dimension
        public let z: Float // Z dimension

        init(_ r: BinaryReader, _ dataSize: Int) {
            x = r.readLESingle()
            y = r.readLESingle()
            z = r.readLESingle()
        }
    }

    public override var description: String { return "SBSP: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var DNAM: DNAMField!
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "DNAM": DNAM = DNAMField(r, dataSize)
        default: return false
        }
        return true
    }
}
